File: Artemis2MenuBar/TrajectorySimulator.swift
import Foundation

/// Simulates mission trajectory using waypoints from the config.
class TrajectorySimulator {

    private let launchDate: Date
    private let earthRadius: Double
    private let moonRadius: Double
    private let waypoints: [MissionConfig.Waypoint]
    private let phases: [MissionConfig.Phase]

    init(config: MissionConfig) {
        launchDate = config.launchDate ?? Date()
        earthRadius = config.trajectory.earthRadius
        moonRadius = config.trajectory.moonRadius
        waypoints = config.trajectory.waypoints
        phases = config.phases
    }

    func getData(at date: Date) -> MissionData {
        let met = date.timeIntervalSince(launchDate)
        return getData(met: met)
    }

    func getData(met: TimeInterval) -> MissionData {
        var data = MissionData()
        data.missionElapsedTime = max(0, met)

        guard !waypoints.isEmpty else { return data }

        if met <= waypoints.first!.met {
            let wp = waypoints.first!
            data.distanceFromEarth = max(0, wp.distanceFromEarth - earthRadius)
            data.distanceFromMoon = max(0, wp.distanceFromMoon - moonRadius)
            data.speed = wp.speed
            data.phase = resolvePhase(wp.phase)
            return data
        }

        if met >= waypoints.last!.met {
            let wp = waypoints.last!
            data.distanceFromEarth = max(0, wp.distanceFromEarth - earthRadius)
            data.distanceFromMoon = max(0, wp.distanceFromMoon - moonRadius)
            data.speed = wp.speed
            data.phase = resolvePhase(wp.phase)
            return data
        }

        var lowerIndex = 0
        for i in 0..<waypoints.count - 1 {
            if met >= waypoints[i].met && met < waypoints[i + 1].met {
                lowerIndex = i
                break
            }
        }

        let wp0 = waypoints[lowerIndex]
        let wp1 = waypoints[lowerIndex + 1]
        let dt = wp1.met - wp0.met
        let t = (met - wp0.met) / dt
        let s = smoothstep(t)

        data.distanceFromEarth = max(0, lerp(wp0.distanceFromEarth, wp1.distanceFromEarth, s) - earthRadius)
        data.distanceFromMoon = max(0, lerp(wp0.distanceFromMoon, wp1.distanceFromMoon, s) - moonRadius)
        data.speed = lerp(wp0.speed, wp1.speed, s)
        data.phase = resolvePhase(t < 0.5 ? wp0.phase : wp1.phase)

        return data
    }

    private func resolvePhase(_ id: String) -> MissionPhaseInfo {
        if let p = phases.first(where: { $0.id == id }) {
            return MissionPhaseInfo(id: p.id, name: p.name, description: p.description, icon: p.icon)
        }
        return MissionPhaseInfo(id: id, name: id, description: "", icon: "questionmark.circle")
    }

    private func lerp(_ a: Double, _ b: Double, _ t: Double) -> Double {
        a + (b - a) * t
    }

    private func smoothstep(_ t: Double) -> Double {
        let c = min(max(t, 0), 1)
        return c * c * (3 - 2 * c)
    }
}

File: OrbitalWindows/Services/TrajectorySimulator.cs
using Orbital.Models;

namespace Orbital.Services;

public class TrajectorySimulator
{
    private readonly DateTime _launchDate;
    private readonly double _earthRadius;
    private readonly double _moonRadius;
    private readonly List<Waypoint> _waypoints;
    private readonly List<Phase> _phases;

    public TrajectorySimulator(MissionConfig config)
    {
        _launchDate = config.LaunchDate ?? DateTime.UtcNow;
        _earthRadius = config.TrajectoryInfo.EarthRadius;
        _moonRadius = config.TrajectoryInfo.MoonRadius;
        _waypoints = config.TrajectoryInfo.Waypoints;
        _phases = config.Phases;
    }

    public MissionData GetData(DateTime date)
    {
        var met = (date - _launchDate).TotalSeconds;
        return GetData(met);
    }

    public MissionData GetData(double met)
    {
        var data = new MissionData
        {
            MissionElapsedTime = Math.Max(0, met)
        };

        if (_waypoints.Count == 0) return data;

        if (met <= _waypoints[0].Met)
        {
            var wp = _waypoints[0];
            data.DistanceFromEarth = Math.Max(0, wp.DistanceFromEarth - _earthRadius);
            data.DistanceFromMoon = Math.Max(0, wp.DistanceFromMoon - _moonRadius);
            data.Speed = wp.Speed;
            data.Phase = ResolvePhase(wp.Phase);
            return data;
        }

        if (met >= _waypoints[^1].Met)
        {
            var wp = _waypoints[^1];
            data.DistanceFromEarth = Math.Max(0, wp.DistanceFromEarth - _earthRadius);
            data.DistanceFromMoon = Math.Max(0, wp.DistanceFromMoon - _moonRadius);
            data.Speed = wp.Speed;
            data.Phase = ResolvePhase(wp.Phase);
            return data;
        }

        var lowerIndex = 0;
        for (var i = 0; i < _waypoints.Count - 1; i++)
        {
            if (met >= _waypoints[i].Met && met < _waypoints[i + 1].Met)
            {
                lowerIndex = i;
                break;
            }
        }

        var wp0 = _waypoints[lowerIndex];
        var wp1 = _waypoints[lowerIndex + 1];
        var dt = wp1.Met - wp0.Met;
        var t = (met - wp0.Met) / dt;
        var s = Smoothstep(t);

        data.DistanceFromEarth = Math.Max(0, Lerp(wp0.DistanceFromEarth, wp1.DistanceFromEarth, s) - _earthRadius);
        data.DistanceFromMoon = Math.Max(0, Lerp(wp0.DistanceFromMoon, wp1.DistanceFromMoon, s) - _moonRadius);
        data.Speed = Lerp(wp0.Speed, wp1.Speed, s);
        data.Phase = ResolvePhase(t < 0.5 ? wp0.Phase : wp1.Phase);

        return data;
    }

    private MissionPhaseInfo ResolvePhase(string id)
    {
        var p = _phases.FirstOrDefault(p => p.Id == id);
        if (p != null)
            return new MissionPhaseInfo { Id = p.Id, Name = p.Name, Description = p.Description, Icon = p.Icon };
        return new MissionPhaseInfo { Id = id, Name = id, Description = "", Icon = "questionmark.circle" };
    }

    private static double Lerp(double a, double b, double t) => a + (b - a) * t;

    private static double Smoothstep(double t)
    {
        var c = Math.Clamp(t, 0, 1);
        return c * c * (3 - 2 * c);
    }
}
